src/mawpy/io.py

```diff
@@ -0,0 +1,119 @@
+import fsspec
+import os
+import pandas as pd
+import importlib
+from typing import Callable
+
+FUNCTION = "func"
+EXTENSION = "ext"
+
+supported_files = {
+    "csv": {FUNCTION: "pandas.read_csv", EXTENSION: ".csv"},
+    "excel": {FUNCTION: "pandas.read_excel", EXTENSION: ".xlsx"},
+}
+
+
+def _import_func(fqp: str):
+    """Take a fully-qualified path and return the imported function.
+
+    ``fqp`` is of the form "package.module.func" or
+    "package.module:subobject.func".
+
+    Warnings
+    --------
+    This can import arbitrary modules. Make sure you haven't installed any modules
+    that may execute malicious code at import time.
+    """
+    if ":" in fqp:
+        mod, name = fqp.rsplit(":", 1)
+    else:
+        mod, name = fqp.rsplit(".", 1)
+
+    mod = importlib.import_module(mod)
+    for part in name.split("."):
+        mod = getattr(mod, part)
+
+    if not isinstance(mod, Callable):
+        raise TypeError(f"{fqp} is not a function")
+
+    return mod
+
+
+def _get_pandas_reader(file_path: str) -> Callable:
+    """
+    Takes an input file path and returns
+    the appropriate pandas reader function from
+    the ``supported_files`` dictionary
+
+    Parameters
+    ----------
+    file_path : str
+        The path file string
+
+    Returns
+    -------
+    Callable
+        The pandas reader function
+
+    Raises
+    ------
+    NotImplementedError
+        If the file extension is not supported
+    """
+    # Get the file base name, e.g. ``file.csv``
+    # and extract the extension, e.g. ``.csv``
+    file_name = os.path.basename(file_path)
+    _, ext = os.path.splitext(file_name)
+    try:
+        # Get the proper supported file format dictionary
+        format_dict = next(
+            format_dict
+            for format_dict in supported_files.values()
+            if format_dict.get(EXTENSION) == ext
+        )
+        pd_class = format_dict.get(FUNCTION)
```

Comment:
I think this would be simpler if it just did the dict lookup, eg:
```
func_name = format_dict.get(ext)
if not func_name:
    raise ...
```

Reply:
I'd like it to be simpler, but I made it a little more complex because excel extension is not `.excel`.. I suppose I can make the keys `.csv` or `.xlsx`.. but then older excel might be a `.xls` extension so then technically the extension field would be a list. 

Also, I didn't do direct import until the function is actually used, it's essentially delaying the import.